### PR TITLE
display okta user and org in password prompt / failed auth error message

### PR DESCRIPTION
**Pull request by gord5500**
### Why

Potential for the config file to have bad user / org values and it's not immediately obvious to the user that those could be the reason for their okta auth failing. This can happen with our internal install script when running wsl and the user gets set to 'root' in the config file.

image

### Ticket

VUL-330

### CODE REVIEW

File: src/D2L.Bmx/ParameterDescriptions.cs [hunk 1]
```diff
@@ -3,7 +3,7 @@ namespace D2L.Bmx;
 internal static class ParameterDescriptions {
 	public const string Org = "Okta org short name or domain name";
 	public const string User = "Okta username";
-	public const string Password = "Okta password";
+	public const string Password = "Okta password for username '{0}' in org '{1}'";
```
Comment (gord5500) [inline src/D2L.Bmx/ParameterDescriptions.cs:6]: The wording to include org felt a little weird to come up with

Comment (cfbao) [inline src/D2L.Bmx/ParameterDescriptions.cs:6]: `ParameterDescriptions.Password` should remain unchanged. It's meant to be succinct description of the "password" parameter itself.
It's also used in the `--help` message where it really doesn't make sense to show it like this (including org & username).

File: src/D2L.Bmx/ConsolePrompter.cs [hunk 2]
```diff
@@ -84,7 +84,8 @@ internal class ConsolePrompter : IConsolePrompter {
 			readKey = () => (char)_stdinReader.Read();
 		}
 
-		Console.Error.Write( $"{ParameterDescriptions.Password}: " );
+		string passwordPrompt = string.Format( ParameterDescriptions.Password, user, org );
+		Console.Error.Write( $"{passwordPrompt}: " );
```
Comment (cfbao) [inline src/D2L.Bmx/ConsolePrompter.cs:88]: A single line is getting long. Not super readable at a glance.

Let's do it like this
```csharp
Console.Error.Write( $"{ParameterDescriptions.Org}: {org}" );
Console.Error.Write( $"{ParameterDescriptions.User}: {user}" );
Console.Error.Write( $"{ParameterDescriptions.Password}: " );
```

This also keeps all parameter descriptions unchanged.

**Review by cfbao — CHANGES_REQUESTED** [review]

File: src/D2L.Bmx/ConsolePrompter.cs (hunk 2 above)
Comment (gord5500) [reply src/D2L.Bmx/ConsolePrompter.cs:88]: > Let's do it like this

Sure

File: src/D2L.Bmx/ParameterDescriptions.cs (hunk 1 above)
Comment (gord5500) [reply src/D2L.Bmx/ParameterDescriptions.cs:6]: Ah okay I didn't think about it showing up in `--help`

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 3]
```diff
@@ -23,15 +23,18 @@ internal interface IOktaApi {
 internal class OktaApi : IOktaApi {
 	private readonly CookieContainer _cookieContainer;
 	private readonly HttpClient _httpClient;
+	private string organization = string.Empty;
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:26]: Please don't initialize things to invalid non-null values just so we can mark things as non-nullable and skip null checks.
Nullables are good. Forced null checks are good. That's the compiler helping us avoid null errors.
False non-null values are much worse.

also - we name private fields with an underscore
```suggestion
	private string? _organization;
```

File: src/D2L.Bmx/ParameterDescriptions.cs (hunk 1 above)
Comment (cfbao) [reply src/D2L.Bmx/ParameterDescriptions.cs:6]: should check all usages (find references) before such a change

File: src/D2L.Bmx/Okta/OktaApi.cs (hunk 3 above)
Comment (gord5500) [reply src/D2L.Bmx/Okta/OktaApi.cs]: Yea that's fair. I saw we couldn't reach the exception without having that value set already but I can see the reasoning

Comment (cfbao) [reply src/D2L.Bmx/Okta/OktaApi.cs]: related - https://desire2learn.atlassian.net/browse/VUL-104 this should help

File: src/D2L.Bmx/Okta/OktaApi.cs [hunk 4]
```diff
@@ -83,7 +86,11 @@ internal class OktaApi : IOktaApi {
 				authnResponse.Embedded.Factors
 			);
 		}
-		throw new BmxException( "Okta authentication failed. Check if org, user and password is correct" );
+
+		string org = _organization is not null ? _organization : "unknown";
```
Comment (cfbao) [inline src/D2L.Bmx/Okta/OktaApi.cs:90]:
```suggestion
		string org = _organization ?? "unknown";
```

**Review by cfbao — APPROVED** [review]

**Review by cfbao — APPROVED** [review]

**Pull request merged**